tooling/greybox_fuzzer/src/lib.rs

```diff
@@ -0,0 +1,1356 @@
+
+        Self {
+            acir_program: acir_and_brillig_programs.acir_program,
+            brillig_program: acir_and_brillig_programs.brillig_program,
+            acir_executor,
+            brillig_executor,
+            location_to_feature_map,
+            brillig_coverage_ranges,
+            mutator,
+            package_name: package_name.to_string(),
+            function_name: function_name.to_string(),
+            num_threads: fuzz_execution_config.num_threads,
+            failure_configuration,
+            corpus_dir: PathBuf::from(
+                folder_configuration.corpus_dir.unwrap_or(DEFAULT_CORPUS_FOLDER.to_string()),
+            ),
+            minimize_corpus: folder_configuration.minimized_corpus_dir.is_some(),
+            minimized_corpus_dir: PathBuf::from(
+                folder_configuration.minimized_corpus_dir.unwrap_or_default(),
+            ),
+            timeout: fuzz_execution_config.timeout,
+            metrics: Metrics::default(),
+        }
+    }
+
+    /// Given the witness from ACIR execution and coverage from Brillig execution, check if they have any new coverage
+    /// ACIR witness is optional, since we can skip ACIR execution or it could have failed, but we collected coverage from brillig
+    /// We never expect brillig to have no coverage (at least for now)
+    fn detect_new_coverage_from_witness_and_brillig(
+        accumulated_coverage: &AccumulatedFuzzerCoverage,
+        witness: &Option<WitnessStack<FieldElement>>,
+        brillig_coverage: &Option<RawBrilligCoverage>,
+    ) -> bool {
+        // Get a list of boolean witnesses (default or taken from accumulated coverage)
+        let mut default_list = PotentialBoolWitnessList::default();
+        let mut bool_witness_list =
+            accumulated_coverage.potential_bool_witness_list.as_ref().unwrap_or(&default_list);
+
+        // If ACVM solved the witness, collect boolean states
+        if witness.is_some() {
+            default_list = accumulated_coverage
+                .potential_bool_witness_list
+                .as_ref()
+                .unwrap_or(&default_list)
+                .merge_new(witness.as_ref().unwrap());
+            bool_witness_list = &default_list;
+        }
+
+        // Form a coverage object with coverage from this run
+        // We don't care about the testcase id, since we are not merging this, just detecting if it has new coverage
+        let new_coverage = SingleTestCaseCoverage::new(
+            TestCaseId::default(),
+            witness,
+            brillig_coverage.clone().unwrap(),
+            bool_witness_list,
+        );
+        // Quickly detect if there is any new coverage so that later single-threaded check can quickly discard this testcase
+        accumulated_coverage.detect_new_coverage(&new_coverage)
+    }
+
+    /// Start the fuzzing campaign
+    pub fn fuzz(&mut self) -> FuzzTestResult {
```

Comment:
It would be nice if it were possible to break this method down into using some helper functions, mainly because it's very long and also to turn individual processing chunks into separate functions.

Reply:
This is already broken up, but I'll try